src/Microsoft.Restier.Core/Conventions/ConventionBasedApiModelBuilder.cs

```diff
@@ -56,11 +49,14 @@ private ConventionBasedApiModelBuilder(Type targetType)
             Ensure.NotNull(builder, "builder");
             Ensure.NotNull(targetType, "targetType");
 
-            var provider = new ConventionBasedApiModelBuilder(targetType);
-            builder.AddHookHandler<IModelBuilder>(provider);
-            builder.AddHookHandler<IModelMapper>(provider);
-            builder.AddHookHandler<IQueryExpressionExpander>(provider);
-            builder.AddHookHandler<IQueryExpressionSourcer>(provider);
+            // The model builder must maintain a singleton life time, for holding states and being injected into
+            // some other services.
+            builder.AddInstance(new ConventionBasedApiModelBuilder(targetType));
+
+            builder.ChainPrevious<IModelBuilder, ConventionBasedApiModelBuilder>();
```

Comment:
Do you mean to split the two functionalities (`IModelBuilder` and `IModelMapper`) of `ConventionBasedApiModelBuilder`?